File: src/main/java/org/nyxlang/interpreter/visitor/VarDeclareVisitor.java
package org.nyxlang.interpreter.visitor;

import org.nyxlang.interpreter.IInterpreter;
import org.nyxlang.parser.ast.INode;
import org.nyxlang.parser.ast.VarDeclaration;
import org.nyxlang.parser.ast.VarDeclareNode;


/**
 * Interprets the var declaration node.
 */
public class VarDeclareVisitor implements IVisitor {

    @Override
    public boolean matches(INode node) {
        return VarDeclareNode.class == node.getClass();
    }

    @Override
    public String visit(IInterpreter interpreter, INode node) {
        var varDeclarationNode = (VarDeclareNode) node;
        for (VarDeclaration declaration : varDeclarationNode.getDeclarations()) {
            interpreter.evalNode(declaration.getTypeExpr());
            interpreter.getGlobalMemory().set(
                    declaration.getIdentifier(),
                    interpreter.evalNode(declaration.getAssignmentExpr()));
        }
        return "<var create>";
    }
}

File: src/main/java/org/nyxlang/interpreter/visitor/VarDeclareVisitor.kt
package org.nyxlang.interpreter.visitor

import org.nyxlang.interpreter.IInterpreter
import org.nyxlang.interpreter.result.IRuntimeResult
import org.nyxlang.interpreter.result.emptyResult
import org.nyxlang.parser.ast.INode
import org.nyxlang.parser.ast.VarDeclareNode

/**
 * Interprets the var declaration node.
 */
class VarDeclareVisitor : IVisitor {
    override fun matches(node: INode) = VarDeclareNode::class == node::class
    override fun visit(interpreter: IInterpreter, node: INode): IRuntimeResult {
        val varDeclarationNode = node as VarDeclareNode
        for (declaration in varDeclarationNode.declarations) {
            val name = declaration.identifier
            val type = interpreter.evalNode(declaration.typeExpr).data
            val value = interpreter.evalNode(declaration.assignmentExpr).data
            interpreter.activationRecord!!.define(name, value)
        }
        return emptyResult()
    }
}